compiler/util/exprAnalysis.cpp

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright 2004-2016 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "astutil.h"
+#include "baseAST.h"
+#include "stlUtil.h"
+#include "stmt.h"
+#include "exprAnalysis.h"
+
+std::map<Expr*,bool> safeExprCache;
+std::map<FnSymbol*,bool> globalManipFuncCache;
+
+/*
+ * Returns true if `e` has no side effects. Checked side effects are:
+ *  - Read/write to a global
+ *  - Is/containsessential primitive
+ *  - If it's a function call has ref arguments
+ */
+bool exprHasNoSideEffects(Expr* e) {
+  if(safeExprCache.count(e) > 0) {
+    return safeExprCache[e];
```

Comment:
I would think that this safeExprCache is actually only really useful when denormalize is thinking about a particular function (for example). Calls to other functions would be handled by the cache of functions (globalManipFuncCache), right?
That suggests to me an different interface in exprAnalysis.h... e.g. two "analysis" - is a function side effect free? and is an expression side effect free? that each have different state.